image_writers_guide/guidelines.adoc

```diff
@@ -9,4 +9,85 @@
+
+For more information on how Volumes are used in OpenShift, see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/volumes.md[this documentation]. (NOTE to docs team:  this link should really go to something in the openshift docs, once we have it)
+
+Keep in mind that even with persisted volumes, each instance of your image will have its own volume, the filesystem is not shared between instances.  This means the volume cannot be used to share state in a cluster.
+
+=== Use Services for inter-image communication
+For cases where your image needs to communicate with a service provided by another image (e.g. a web frontend image that needs to access a database image to store/retrieve data), your image should consume an OpenShift Service.  Services provide a static endpoint for access which will not change as containers are stopped/started/moved.  In addition services provide load balancing for requests.  For more information see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/services.md[this documentation].  (NOTE to docs team:  this link should really go to something in the openshift docs once we have it)
+
+=== Provide Common Libraries
+Ensure that your image contains commonly used libraries for your platform.  In particular, provide database drivers for common databases used with your platform (e.g. provide JDBC drivers for MySQL and PostgreSQL if you are creating a java framework image).  This will prevent the need for common dependencies to be downloaded during application assembly time, speeding up application image builds.  It will also simplify the work application authors need to do to ensure all their dependencies are met.
+
+=== Configuration
+Runtime configuration should be handled via environment variables.  This means that configuration files should be parameterized with environment variables and the values should be substituted in during image startup.  This allows consumers of your image to customize behavior (e.g. database settings, passwords, performance tuning) without having to introduce a new layer on top of your image.  Instead, they can simply define environment variable values when defining a pod, and change those settings without rebuilding the image.
+
+This topic ties in with using services for inter-image communication in that configuration like datasources should be defined in terms of environment variables which provide the service endpoint information.  In this way, an application can dynamically consume a datasource service that is defined in the OpenShift environment without modifying the application image.
+
+In addition, tuning should be done by inspecting the cgroups settings for the container.  This will allow the image to tune itself to the available memory, CPU, etc.  For example java based images should tune their heap based on the cgroup maximum memory parameter to ensure they do not exceed the limits and get an out of memory error.
```

Comment:
i've added some doc about cgroups, but i can't actually find anything useful that shows how to read cgroups info from within the container, only from the host.  this may need to remain a TBD for now.